Modal: Add small top padding to the content so that avoid cutting off the visible outline when hovering items

t-hamano opened the pull request:
_Update:_

This PR originally tried to fix the cutoff when focusing on an item in the "Add template" modal. However, with #51806, the text now appears above the item and I have confirmed that this problem no longer occours.

However, the same problem occurs when there is a focusable element at the beginning of the modal content. Therefore, I decided to add 2px top padding to the content of the modal component itself.

## What?
This PR adds 2px top padding to the content of the `Modal` component. This prevents the problem of the outline being cut off in advance when a focusable element is inserted into the modal content.

![modal_cutoff](https://github.com/WordPress/gutenberg/assets/54422211/b1297d2c-516e-4bb5-b1e1-d26bd3872073)

## How?
Use the `$border-width-focus-fallback` variable to add 2px padding to the top of the modal content.

## Testing Instructions

### Rename Navigation Modal

- Access Site Editor > Navigation > Click ellipsis button > Select "Rename".
- This PR removed the top padding of the rename modal form, but thanks to the top padding of the modal component itself, the focus outline should not cut off as before.

![navigation-rename-modal](https://github.com/WordPress/gutenberg/assets/54422211/76c84692-fd89-4928-ab05-7c6fa07c34cb)

### Choose Pattern Modal

I have found that adding top padding to the Modal component itself fixes the outline cut-off in the following two existing modals:

- The Choose Pattern Modal displayed when creating a custom template
- The Choose Pattern Modal displayed when creating a new post

![choose a pattern](https://github.com/WordPress/gutenberg/assets/54422211/94128333-2653-4e49-b87c-fe8337384a2a)

These modals [already have 2px top padding to prevent outline cutting on focus](https://github.com/WordPress/gutenberg/blob/a3af2c7d0aee82a0cb8f74beae95730849b5afc6/packages/edit-site/src/components/start-template-options/style.scss#L32-L33). However, when focused, the outlines are [extended by an additional 2px due to the offset](https://github.com/WordPress/gutenberg/blob/a3af2c7d0aee82a0cb8f74beae95730849b5afc6/packages/block-editor/src/components/block-patterns-list/style.scss#L39-L41), resulting in the outlines being cut off. By adding top padding to the modal component itself, it now has a total of 4px top padding, which also fixes the outline cutoff on focus.

t-hamano committed:
Site Editor: Fix focus cutoff in add template modal

t-hamano commented:
Since Gutenberg 16.1 has already been released, I have added the backport label just in case.

jasmussen commented:
Relevant: #51806. CC: @jameskoster

jasmussen commented:
Is there any way we can fix this in the modal component itself? I.e. right now there's a little bottom padding baked into the modal header:

<img width="888" alt="Screenshot 2023-06-23 at 10 25 03" src="https://github.com/WordPress/gutenberg/assets/1204802/15e25c0a-34b3-49ea-907a-3a559355427b">

But there's no top padding baked into the content area:

<img width="975" alt="Screenshot 2023-06-23 at 10 25 11" src="https://github.com/WordPress/gutenberg/assets/1204802/7058f0a2-9cbf-4699-a0e7-6ede5e9c13d0">

Could we split that padding in two, have half be bottom padding on the header, and the other half be top padding on the content?

jameskoster commented:
If we split the padding then the stroke which appears beneath the header on scroll might look a bit close to the title text.

Adding 4px padding-top to `.components-modal__content` could be worth a try?

t-hamano commented:
This problem will always occur when a focusable element, such as a button, textbox, etc., is placed at the beginning of the modal content. Therefore, it might be a good idea to add slight padding to the top of the modal content.

@ciampo

Is there any problem with the approach of adding a slight padding at the top of the `.components-modal__content` to prevent the cutoff of the focus outline?

![modal_cutoff](https://github.com/WordPress/gutenberg/assets/54422211/b1297d2c-516e-4bb5-b1e1-d26bd3872073)

ciampo commented:
I understand the problem, but we shouldn't be using internal components classnames (like `.components-modal__content`) outside of the components package.

Couldn't we add small `margin-top` on the `edit-site-add-new-template__template-list__contents` grid container? Alternatively, could we change the `Modal` component to always have a small padding on its top container? In both cases, `2px` would be enough to avoid cutting off the focus ring

t-hamano commented:
Thanks for the advice, @ciampo.

> Couldn't we add small margin-top on the edit-site-add-new-template__template-list__contents grid container? Alternatively, could we change the Modal component to always have a small padding on its top container? In both cases, 2px would be enough to avoid cutting off the focus ring

I would like to try the latter method so that we don't have to add `padding-top` every time we encounter similar problems in the future.

t-hamano committed:
Add padding-top to the modal content

t-hamano committed:
Remove unnecessary padding-top

t-hamano committed:
Remove unnecessary padding-top

t-hamano committed:
Update changelog

t-hamano committed:
Revert top padding from block patterns list

t-hamano committed:
Revert top padding from block patterns list

t-hamano committed:
Remove unnecessary changes

t-hamano committed:
Remove unnecessary changes

t-hamano commented:
I have changed this PR as follows:

- Added 2px top padding to the content of the modal component itself. This should cause no outline cutoff when the element with the outline appears at the beginning of the content when in focus.
- The top padding of the navigation rename modal has been removed as it is no longer needed.

As a result, the target to be fixed by this PR has changed. Please see details in the description of the updated PR.

ciampo commented on packages/components/src/modal/style.scss:
```diff
@@ -130,7 +130,7 @@
 	flex: 1;
 	margin-top: $header-height + $grid-unit-15;
-	padding: 0 $grid-unit-40 $grid-unit-40;
+	padding: $border-width-focus-fallback $grid-unit-40 $grid-unit-40;
```
It may be worth to add a short inline comment explaining the reason for the top padding

ciampo reviewed (approved):
LGTM, although I'll let Jay/Joen give the final green light

t-hamano committed:
Add CSS inline comment

jameskoster commented:
I would be tempted to use 4px, so that the overall dimensions continue to align with the 4px baseline grid.

4px would also account for more elaborate focus styles, such as pattern previews, which have an effective 3px outline (1.5px white + 1.5px admin theme color):

<img width="773" alt="Screenshot 2023-06-26 at 09 38 22" src="https://github.com/WordPress/gutenberg/assets/846565/65c3b0aa-14a4-4403-9ce2-a2093827c908">

jasmussen commented:
This seems to solve the issue in a generic way. Top padding is the same:

<img width="856" alt="Screenshot 2023-06-26 at 10 38 06" src="https://github.com/WordPress/gutenberg/assets/1204802/67ddb45e-03b8-415e-b503-255faa33069b">

Content padding is now 2px at the top:

<img width="838" alt="Screenshot 2023-06-26 at 10 37 54" src="https://github.com/WordPress/gutenberg/assets/1204802/d9dc698c-821a-466d-bc40-203225f620f5">

However if I don't misread this, that means the spacing between the header title and the body content is now 2px taller, which puts the metrics off a little bit. 

So to simplify the whole thing, I would:

* Apply the 4px padding at the top of the content part (`$grid-unit-05`)
* _Subtract_ those same 4px from the bottom padding of the header 

So for the header instead of `padding: $grid-unit-30 $grid-unit-40 $grid-unit-15;`, you'd have something like `padding: ($grid-unit-30 - $grid-unit-05) $grid-unit-40 $grid-unit-15;`

t-hamano commented:
@jasmussen

> * _Subtract_ those same 4px from the bottom padding of the header 

The `Modal` component draws a border at the bottom of the header if the content is scrollable. Therefore, reducing the bottom padding of the modal header by 4px will cause the header text to drop slightly

### Bottom padding is 12px

![before](https://github.com/WordPress/gutenberg/assets/54422211/24ee6e9e-a6df-4a60-87d2-2b895d9bc150)

### Bottom padding is 8px

![after](https://github.com/WordPress/gutenberg/assets/54422211/960f8594-20b1-40ed-92d7-ef87be1419e9)

The approach of changing the top padding of the modal content from 2px to 4px seems to have less visual impact. What do you think?

jasmussen commented:
It's a bit hard to discern the difference between those two screenshots, which to me suggests it's okay for that text to drop a bit. 

The default metrics should still ideally be somewhat close to this grid:
<img width="1053" alt="Screenshot 2023-06-26 at 15 40 15" src="https://github.com/WordPress/gutenberg/assets/1204802/ba7b8278-0e3d-479b-9334-83e0bea70646">

Hope that helps!

t-hamano committed:
Merge branch 'trunk' into fix/add-template-cutoff

t-hamano committed:
Change padding metrics

t-hamano commented:
I have made the following changes:

- Top padding of the modal content to **4px**
- Bottom padding of the modal header from **8px** to **4px**
- Top padding, no longer needed in the pattern selection modal, has also been removed

This looks fine to me. Below are some screenshots.

<details><summary>Preferences</summary>

![preferences](https://github.com/WordPress/gutenberg/assets/54422211/b9394739-e466-401e-ae0c-a5a5fde8592b)

</details> 

<details><summary>Keyboard shotycuts (Scrollable)</summary>

![keyboard-shortcuts](https://github.com/WordPress/gutenberg/assets/54422211/2cc5dfb5-dbf7-4e52-9e80-575342a225b7)

</details> 


<details><summary>Rename Navigation</summary>

![navigation-rename](https://github.com/WordPress/gutenberg/assets/54422211/f99586f9-2cb5-4f17-b103-d4b7d08ee635)

</details> 


<details><summary>Choose Pattern Modal displayed when creating a custom template</summary>

![choose-pattern-new-template](https://github.com/WordPress/gutenberg/assets/54422211/0165739b-b78d-4a56-a5cd-6b077c33fc29)

</details> 

<details><summary>Choose Pattern Modal displayed when creating a new post</summary>

![choose-pattern-new-post](https://github.com/WordPress/gutenberg/assets/54422211/b9233ce8-8a3f-492b-8f33-184f4ddc2a69)

</details> 

<details><summary>Choose Pattern Modal displayed when creating a new post (mobile)</summary>

![choose-pattern-new-post-mobile](https://github.com/WordPress/gutenberg/assets/54422211/07dfa415-7628-47e4-a560-0506ccc3113a)

</details> 


<details><summary>Rename Navigation (mobile)</summary>

![rename-mobile](https://github.com/WordPress/gutenberg/assets/54422211/53c122b2-3fe7-411e-864d-14ca6ae54b6c)

</details> 

<details><summary>ConfirmDialog (no header)</summary>

![confirm-dialog](https://github.com/WordPress/gutenberg/assets/54422211/30046a02-984d-4571-97bb-8fc0baaa8582)

</details>

t-hamano committed:
Merge branch 'trunk' into fix/add-template-cutoff

jasmussen reviewed (approved):
Thanks for the patience. Looks good:

<img width="722" alt="Screenshot 2023-06-27 at 10 33 12" src="https://github.com/WordPress/gutenberg/assets/1204802/463350a6-16c1-4577-a6dc-08fd16e49f3c">
<img width="831" alt="Screenshot 2023-06-27 at 10 33 16" src="https://github.com/WordPress/gutenberg/assets/1204802/a4a69dde-da90-476e-9d07-83b39ad72b74">

The preferences dialog looks a little custom, and does not have the top padding, the tabs even misalign:
<img width="904" alt="Screenshot 2023-06-27 at 10 32 37" src="https://github.com/WordPress/gutenberg/assets/1204802/d0287624-92da-4af9-8a22-d797b478fb1c">

However that's also in trunk, so separate:
<img width="921" alt="Screenshot 2023-06-27 at 10 33 48" src="https://github.com/WordPress/gutenberg/assets/1204802/768f4ca7-cd78-41d0-a37f-59930e3e5e70">

Nice work, thank you!

tellthemachines commented:
I just cherry-picked this PR to the add/pre-beta1-bugfixes branch to get it included in the next release: 4f3599df90

stokesman commented:
I think this caused a little visual regression on Blocks screen of the Edit Post Preferences Modal:

<img width="769" alt="image" src="https://github.com/WordPress/gutenberg/assets/9000376/aff5563c-562b-47db-8c19-c4c23d1f5dc9">

Gap above the sticky elements.
